File: go/hot_100/5_longest-palindromic-substring.go
package main

func MaxLength(i int,s string) (int,int,int){
	single,double:=1,0
	m,n:=i,i+1
	for j:=1;i+j<len(s)&&i-j>=0&&s[i+j]==s[i-j];j++{
		single+=2
		m-=1
		n+=1
	}
	if i!=len(s)-1 && s[i]==s[i+1]{
		double=2
		l,r:=i,i+2
		for j:=1;i+j+1<len(s)&&i-j>=0&&s[i+j+1]==s[i-j];j++{
			double+=2
			l-=1
			r+=1
		}
		if double>single{
			return double,l,r
		}
	}
	
	return single,m,n
}

func longestPalindrome(s string) string {
	length:=0
	var res string
	for i, _ := range s {
		nowLen,x,y:=MaxLength(i,s)
		if nowLen>length{
			res=s[x:y]
			length=nowLen
		}
	}
	return res
}

// func main(){
// 	longestPalindrome("babad")
// }

File: python/hot_100/5_longest-palindromic-substring.py
# class Solution:
#     def longestPalindrome(self, s: str) -> str:
#         def inner(i,j):
#             while i>=0 and j<len(s) and s[i]==s[j]:
#                 i-=1
#                 j+=1
#             return i+1,j-1
#         if len(s)==1:
#             return s
#         max_len=0
#         for i in range(len(s)-1):
#             m,n=inner(i,i)
#             single=n-m+1
#             if single>max_len:
#                 max_len=single
#                 res=s[m:n+1]
#             m,n=inner(i,i+1)
#             double=n-m+1
#             if double>max_len:
#                 max_len=double
#                 res=s[m:n+1]
#         return res
class Solution(object):
    def longestPalindrome(self, s):
        res = ''
        for i in range(len(s)):
            start = max(i - len(res) -1, 0)
            temp = s[start: i+1]
            if temp == temp[::-1]:
                res = temp
            else:
                temp = temp[1:]
                if temp == temp[::-1]:
                    res = temp
        return res

s=Solution()
s.longestPalindrome('cccdecbabced')